imagemagick: add liquid rescale build option

DieracDelta opened the pull request:
###### Motivation for this change
It is not currently possible to build imagemagick with liquid rescaling on nix without an overlay.

###### Things done
This PR adds a optional dependency on liblqr1. If this library exists, imagemagick builds with with liquid rescale support.

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

ajs124 reviewed (approved):
`"--with-lqr=yes"` probably works just as `"--with-lqr"` without the `=yes`.

This seems like it might be useful to some people? Idk how the feature set of packages like this is decided on, tbh.

DieracDelta commented:
You're right, I can definitely remove the --yes part. I ran into this as a feature that I wanted when running imagemagick. Maybe this is easy thing to add in via an overlay? On gentoo, I enabled this with a useflag. Nix doesn't look like it has an equivalent to useflags, though. In any case, it seems worthwhile to take note of it somewhere in case someone else runs into this problem.

gebner commented:
This looks great!  IMHO packages in nixpkgs should always be built with the maximum set of features enabled.  (Because it's hard to substitute e.g. imagemagick with your own version if it is used as a dependency by something else.)

1. This changes causes a lot of rebuilds (because many packages depend on imagemagick).  Please make this PR to `staging` instead of `master`.  (You can edit this PR, no need to make a new one.)
1. Can we also add this dependency to `imagemagick7`?
1. Please add `liblqr1 = null;` to the overrides in the `imagemagick_light` package.  (This is in the `pkgs/top-level/all-packages.nix` file).

SuperSandro2000 commented:
> This changes causes a lot of rebuilds (because many packages depend on imagemagick). Please make this PR to staging instead of master. (You can edit this PR, no need to make a new one.)

100 are fine.

@DieracDelta

DieracDelta commented:
Thanks for the feedback! Per SuperSandro2000's suggestion, I've left this on master. I've made the suggested modifications. Sorry for the long delay on fixing this.

Changes:
- rebased onto head master
- copied changes to imagemagick 7.0 package
- removed `=yes` from `--with-lqr` flag.
- disabled liblqr when building `imagemagick_light` and `imagemagick7_light` packages.

Hopefully this passes CI.

SuperSandro2000 reviewed (approved):
diff LGTM

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review which is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 83149` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>imagemagick</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

imagemagick:
<details>

Please consider this feature to be alpha.

A substituteInPlace with an unused --replace got detected:
```
substituteStream(): WARNING: pattern '/nix/store/pjrkzmdm3p93r2la8j5bvksh5cq84535-pkg-config-wrapper-0.29.2/bin/pkg-config -config' doesn't match anything in file '/nix/store/ji7mdabnkd2ql7808ygbcll4cnmbznd3-imagemagick-6.9.11-29-dev/bin/Magick++-config'
substituteStream(): WARNING: pattern '/nix/store/pjrkzmdm3p93r2la8j5bvksh5cq84535-pkg-config-wrapper-0.29.2/bin/pkg-config' doesn't match anything in file '/nix/store/ji7mdabnkd2ql7808ygbcll4cnmbznd3-imagemagick-6.9.11-29-dev/bin/Magick++-config'
substituteStream(): WARNING: pattern '/nix/store/pjrkzmdm3p93r2la8j5bvksh5cq84535-pkg-config-wrapper-0.29.2/bin/pkg-config -config' doesn't match anything in file '/nix/store/ji7mdabnkd2ql7808ygbcll4cnmbznd3-imagemagick-6.9.11-29-dev/bin/Magick-config'
substituteStream(): WARNING: pattern '/nix/store/pjrkzmdm3p93r2la8j5bvksh5cq84535-pkg-config-wrapper-0.29.2/bin/pkg-config' doesn't match anything in file '/nix/store/ji7mdabnkd2ql7808ygbcll4cnmbznd3-imagemagick-6.9.11-29-dev/bin/Magick-config'
substituteStream(): WARNING: pattern '/nix/store/pjrkzmdm3p93r2la8j5bvksh5cq84535-pkg-config-wrapper-0.29.2/bin/pkg-config -config' doesn't match anything in file '/nix/store/ji7mdabnkd2ql7808ygbcll4cnmbznd3-imagemagick-6.9.11-29-dev/bin/MagickCore-config'
substituteStream(): WARNING: pattern '/nix/store/pjrkzmdm3p93r2la8j5bvksh5cq84535-pkg-config-wrapper-0.29.2/bin/pkg-config' doesn't match anything in file '/nix/store/ji7mdabnkd2ql7808ygbcll4cnmbznd3-imagemagick-6.9.11-29-dev/bin/MagickCore-config'
substituteStream(): WARNING: pattern '/nix/store/pjrkzmdm3p93r2la8j5bvksh5cq84535-pkg-config-wrapper-0.29.2/bin/pkg-config -config' doesn't match anything in file '/nix/store/ji7mdabnkd2ql7808ygbcll4cnmbznd3-imagemagick-6.9.11-29-dev/bin/MagickWand-config'
substituteStream(): WARNING: pattern '/nix/store/pjrkzmdm3p93r2la8j5bvksh5cq84535-pkg-config-wrapper-0.29.2/bin/pkg-config' doesn't match anything in file '/nix/store/ji7mdabnkd2ql7808ygbcll4cnmbznd3-imagemagick-6.9.11-29-dev/bin/MagickWand-config'
substituteStream(): WARNING: pattern '/nix/store/pjrkzmdm3p93r2la8j5bvksh5cq84535-pkg-config-wrapper-0.29.2/bin/pkg-config -config' doesn't match anything in file '/nix/store/ji7mdabnkd2ql7808ygbcll4cnmbznd3-imagemagick-6.9.11-29-dev/bin/Wand-config'
substituteStream(): WARNING: pattern '/nix/store/pjrkzmdm3p93r2la8j5bvksh5cq84535-pkg-config-wrapper-0.29.2/bin/pkg-config' doesn't match anything in file '/nix/store/ji7mdabnkd2ql7808ygbcll4cnmbznd3-imagemagick-6.9.11-29-dev/bin/Wand-config'
```

Please check the offending substituteInPlace for typos or changes in source.
</details>

DieracDelta commented:
Hmm... not really sure what to make of this error. I'm not able to reproduce it.

To try and reproduce, I cloned my branch of nixpkgs, changed `NIXPKGS` to point to it, then built `imagemagick`, `imagemagick7`, `imagemagick7_light`, and `imagemagick_light`. All built successfully. I also checked the outputs, and the list of features compiled in matched what I would expect. E.g. lqr for the non-light builds, and no lqr for the light builds.

@SuperSandro2000 any idea how to reproduce this error?

SuperSandro2000 commented:
> Hmm... not really sure what to make of this error. I'm not able to reproduce it.

This is not an error but additional feedback.

> It's worth noting I'm only on x86-64. Perhaps the bot built it on another arch?

Only building on amd64.

When you take a look at https://github.com/DieracDelta/nixpkgs/blob/fix/imagemagick-default/pkgs/applications/graphics/ImageMagick/default.nix#L89 you will see the lines mentioned in the comment. They are not actually applied. Maybe they need an update or are outdated and can be removed.

tomberek commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 83149` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
6 packages built:
</summary>
<ul>
<li>imagemagick</li>
<li>imagemagick7</li>
<li>imagemagick7Big</li>
<li>imagemagick7_light</li>
<li>imagemagickBig</li>
<li>imagemagick_light</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
imagemagick:
<details>

Please consider this feature to be alpha.

A substituteInPlace with an unmatched pattern got detected:
```
substituteStream(): WARNING: pattern '/nix/store/vlggxvpr2zfv54q89fs429nqrhzlngb4-pkg-config-wrapper-0.29.2/bin/pkg-config -config' doesn't match anything in file '/nix/store/hlagxfv4360ykxvkvpf2vdml79qg0rf0-imagemagick-6.9.11-60-dev/bin/Magick++-config'
substituteStream(): WARNING: pattern '/nix/store/vlggxvpr2zfv54q89fs429nqrhzlngb4-pkg-config-wrapper-0.29.2/bin/pkg-config' doesn't match anything in file '/nix/store/hlagxfv4360ykxvkvpf2vdml79qg0rf0-imagemagick-6.9.11-60-dev/bin/Magick++-config'
substituteStream(): WARNING: pattern '/nix/store/vlggxvpr2zfv54q89fs429nqrhzlngb4-pkg-config-wrapper-0.29.2/bin/pkg-config -config' doesn't match anything in file '/nix/store/hlagxfv4360ykxvkvpf2vdml79qg0rf0-imagemagick-6.9.11-60-dev/bin/Magick-config'
substituteStream(): WARNING: pattern '/nix/store/vlggxvpr2zfv54q89fs429nqrhzlngb4-pkg-config-wrapper-0.29.2/bin/pkg-config' doesn't match anything in file '/nix/store/hlagxfv4360ykxvkvpf2vdml79qg0rf0-imagemagick-6.9.11-60-dev/bin/Magick-config'
substituteStream(): WARNING: pattern '/nix/store/vlggxvpr2zfv54q89fs429nqrhzlngb4-pkg-config-wrapper-0.29.2/bin/pkg-config -config' doesn't match anything in file '/nix/store/hlagxfv4360ykxvkvpf2vdml79qg0rf0-imagemagick-6.9.11-60-dev/bin/MagickCore-config'
substituteStream(): WARNING: pattern '/nix/store/vlggxvpr2zfv54q89fs429nqrhzlngb4-pkg-config-wrapper-0.29.2/bin/pkg-config' doesn't match anything in file '/nix/store/hlagxfv4360ykxvkvpf2vdml79qg0rf0-imagemagick-6.9.11-60-dev/bin/MagickCore-config'
substituteStream(): WARNING: pattern '/nix/store/vlggxvpr2zfv54q89fs429nqrhzlngb4-pkg-config-wrapper-0.29.2/bin/pkg-config -config' doesn't match anything in file '/nix/store/hlagxfv4360ykxvkvpf2vdml79qg0rf0-imagemagick-6.9.11-60-dev/bin/MagickWand-config'
substituteStream(): WARNING: pattern '/nix/store/vlggxvpr2zfv54q89fs429nqrhzlngb4-pkg-config-wrapper-0.29.2/bin/pkg-config' doesn't match anything in file '/nix/store/hlagxfv4360ykxvkvpf2vdml79qg0rf0-imagemagick-6.9.11-60-dev/bin/MagickWand-config'
substituteStream(): WARNING: pattern '/nix/store/vlggxvpr2zfv54q89fs429nqrhzlngb4-pkg-config-wrapper-0.29.2/bin/pkg-config -config' doesn't match anything in file '/nix/store/hlagxfv4360ykxvkvpf2vdml79qg0rf0-imagemagick-6.9.11-60-dev/bin/Wand-config'
substituteStream(): WARNING: pattern '/nix/store/vlggxvpr2zfv54q89fs429nqrhzlngb4-pkg-config-wrapper-0.29.2/bin/pkg-config' doesn't match anything in file '/nix/store/hlagxfv4360ykxvkvpf2vdml79qg0rf0-imagemagick-6.9.11-60-dev/bin/Wand-config'
```

Please check the offending substituteInPlace for typos or changes in source.
</details>
</details>

SuperSandro2000 commented:
I forgot this PR.

@DieracDelta Can you rebase it and resolve the merge conflict?

DieracDelta committed:
imagemagick: add liquid rescale build option

DieracDelta commented:
@SuperSandro2000 rebased. Built the packages on my system for a x86-64 linux target, and they seem to work fine. :+1:

SuperSandro2000 commented:
> Note that if the substituteInPlace error shows up again for imagemagick 6, I can delete the substitute in place lines ([this](https://github.com/NixOS/nixpkgs/blob/3b392e8951cef55c2c56e8549107ccd0f1920f0d/pkgs/applications/graphics/ImageMagick/6.x.nix#L68) entire for-loop). I built the package with them deleted without a hitch, and it seems to run. I'm not sure why it's there in the first place, though so there may be unintended side effects. In either case that might be better as a separate PR.

I think those are only doing something when you are consuming imagemagick as a library. It would be better to remove that in a separate PR to get this PR faster merged.

Also when I look at the amount of rebuilds we need to target staging here.

DieracDelta commented:
Switched to staging. (and I did not include that change in this PR, so we should be set there)
